### PR TITLE
feat: relax balancing

### PR DESCRIPTION
**Pull request by metacertain**
### Description

This PR intends to change the selection of balanced connections by relaxing the criteria from selecting a closest peer to selecting a close-enough peer for a pseudo-address

### Related Issue (Optional)
https://github.com/ethersphere/bee/issues/2521

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/2677)

### CODE REVIEW

**Review by istae — APPROVED** [review]
perfect, thanks,.

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -231,43 +231,35 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
...
 
-			closestConnectedPeer, err := closestPeer(k.connectedPeers, pseudoAddr, noopSanctionedPeerFn)
+			// Connect to closest known peer which we haven't tried connecting to recently.
+
+			_, err := nClosePeerInSlice(binConnectedPeers, pseudoAddr, noopSanctionedPeerFn, uint8(i+k.bitSuffixLength+1))
 			if err != nil {
-				if errors.Is(err, topology.ErrNotFound) {
-					break
+				if !errors.Is(err, topology.ErrNotFound) {
+					break // not possible
```
Comment (istae) [inline pkg/topology/kademlia/kademlia.go:244]: technically if there is an error here that's not topology.ErrNotFound, we should return, instead of break.

**Review by istae — APPROVED** [review]

File: pkg/topology/kademlia/kademlia.go [hunk 2]
```diff
@@ -231,43 +231,35 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
...
 
-			// Connect to closest known peer which we haven't tried connecting to recently.
-			closestKnownPeer, err := closestPeerInSlice(binPeers, pseudoAddr, skipPeers)
+			closestKnownPeer, err := nClosePeerInSlice(binPeers, pseudoAddr, skipPeers, uint8(i+k.bitSuffixLength+1))
 			if err != nil {
 				if errors.Is(err, topology.ErrNotFound) {
-					break
+					continue
+				} else {
+					return
```
Comment (istae) [inline pkg/topology/kademlia/kademlia.go:255]: return here?

**Review by istae — APPROVED** [review]

File: pkg/topology/kademlia/kademlia.go (hunk 2 above)
Comment (metacertain) [reply pkg/topology/kademlia/kademlia.go:255]: yes, no other error than topology.ErrNotFound is expected here

File: pkg/topology/kademlia/kademlia.go [hunk 3]
```diff
@@ -231,43 +231,35 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
...
 
-			closestConnectedPeer, err := closestPeer(k.connectedPeers, pseudoAddr, noopSanctionedPeerFn)
+			// Connect to closest known peer which we haven't tried connecting to recently.
+
+			_, err := nClosePeerInSlice(binConnectedPeers, pseudoAddr, noopSanctionedPeerFn, uint8(i+k.bitSuffixLength+1))
 			if err != nil {
-				if errors.Is(err, topology.ErrNotFound) {
-					break
+				if !errors.Is(err, topology.ErrNotFound) {
+					return
```
Comment (acud) [inline pkg/topology/kademlia/kademlia.go:244]: there was not a return here, why the change?

Comment (metacertain) [reply pkg/topology/kademlia/kademlia.go:244]: this function does not return any other error than ErrNotFound, so this is non-consequential, I can remove this

Comment (acud) [reply pkg/topology/kademlia/kademlia.go:244]: that's incorrect. there are 2 levels of `for` loops here, the `break` statement means that the immediate enclosing for loop breaks, but the enclosing loop continues

File: pkg/topology/kademlia/kademlia.go (hunk 2 above)
Comment (metacertain) [reply pkg/topology/kademlia/kademlia.go:255]: this return is also non-consequential, it can be removed

Comment (acud) [reply pkg/topology/kademlia/kademlia.go:255]: why return and not `break`? please see the comment on the other change

File: pkg/topology/kademlia/kademlia.go (hunk 3 above)
Comment (metacertain) [reply pkg/topology/kademlia/kademlia.go:244]: the break was in a different context. 
closestpeer to any pseudoaddress always existed if there was at least some connections in the bins.
Now, closeenoughPeer can return a close-enough-peer if there is one, or nothing. 
if there's a close enough connected peer, we skip to next pseudoaddress (continue).
if there is not one, we select one in line 250. 
if there is not one, we get the topology.ErrNotFound error. if any other error is received, probably something is off, so return. this can never happen with the current nClosePeer function

Comment (acud) [reply pkg/topology/kademlia/kademlia.go:244]: interactive review please

**Review by acud — APPROVED** [review]
